README.md
# @sunly95/gen-version

中文 ｜ [English](./README_EN.md)

修改package.json版本号，并生成CHANGELOG.md

<img src="./screenshot.png" width=500 />

## 使用

### 安装

```bash
# npm
npm install @sunly95/gen-version@latest --save-dev

# yarn
yarn add @sunly95/gen-version@latest --dev
```

### 通过命令更新版本

```bash
npx -p @sunly95/gen-version update
```

### 手动调用更新函数更新版本

#### commonJS

```js
// build/index.js
const main = require("@sunly95/gen-version")
main()
```

#### ESModule

```js
// build/index.js
import main from "@sunly95/gen-version"
main()
```

每次调用`build/index.js`时就会自动生成`CHANGELOG.md`并修改`package.json`中的版本号

### 使用脚本配合build命令打包

```
// package.json
{
  "script": {
    "build": "node ./build/index.js && yarn run build",
    "update:version": "npx -p @sunly95/gen-version update"
  }
}
```

之后在打包时使用 `npm run build` 或者 `npm run update:version` 就可以直接调用`build/index.js`并打包

## 配置

运行命令：

```bash
npx -p @sunly95/gen-version init
```

会在项目根目录生成`gen-version.config.json`文件，可以自由配置，默认配置项如下：

```
// gen-version.config.json
{
  title: "CHANGELOG",  // CHANGELOG 标题
  checkVersion: true,  // 检查版本号是否符合 Major.Minor.Patch 的格式
  changelogTemplate: [
    { 新增: "这里记录新增加了哪些功能／接口" },
    { 更改: "这里记录更改了功能／接口变更" },
    { 修复: "这里记录解决了哪些问题" },
    { 废弃: "不建议使用的功能／接口，将来会删掉" },
    { 移除: "之前不建议使用的功能／接口，这次真的删掉了" },
    { 样式: "这里记录了样式的更改" },
    { 类型: "这里记录了类型的更改" },
    { 其他: "这里记录性能优化和安全性增强等改进" }
  ] // CHANGELOG 模板，可以自定义
}
```

## CHANGELOG

[CHANGELOG.md](./CHANGELOG.md)

## LICENSE

[![License: MIT](https://img.shields.io/badge/License-MIT-green.svg)](./LICENSE)


CHANGELOG.md
# CHANGELOG

## [2.3.0] - 2023-08-19

### Feat

- 版本选择

## [2.2.2] - 2023-08-17

### Fix

- 升级版本没有shebang

## [2.2.0] - 2023-08-16

### Feat

- 通过命令升级版本信息

## [2.1.2] - 2023-08-15

### Update

- 使用readline支持低版本node

## [2.1.0] - 2023-05-23

### Feat

- 用户自定义配置

### Fix

- commonJS环境不能引入
- 生成的ChangeLog格式不正确

## [2.0.0] - 2023-05-23

### Feat

- 使用TS重构
- 移除所有第三方依赖

## [1.0.4] - 2023-05-23

### Change

- 删除所有依赖包

## [1.0.3] - 2022-05-10

### Change

- 修改引入方式

## [1.0.1] - 2021-12-08

### Fix

- package.json 没有正确修改

## [1.0.0] - 2021-12-08

### Feat

- 生成版本文件


test/index.ts
/*
 * @Description: 测试文件
 * @Author: Sunly
 * @Date: 2023-05-23 08:12:28
 */
import main from "../src/index.js";

main();


src/test/index.js
/*
 * @Description: test
 * @Author: Sunly
 * @Date: 2021-12-08 15:35:42
 */
const genVersion = require("../index");

genVersion();


src/index.js
/*
 * @Description: update version
 * @Author: Sunly
 * @Date: 2021-12-08 14:02:13
 */
const readline = require("readline");
const fs = require("fs");
const path = require("path");
const { EOL } = require("os");
const packageFile = require(path.resolve(process.cwd(), "./package.json"));

const YEAR = new Date().getFullYear();
const MONTH = (new Date().getMonth() + 1).toString().padStart(2, "0");
const DAY = new Date().getDate().toString().padStart(2, "0");
const TODAY = `${YEAR}-${MONTH}-${DAY}`;

function main() {
  console.log("\n上次构建版本: ", packageFile.version);

  const input = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  let version = "";

  input.question("\n请输入本次构建版本号: ", (ver) => {
    version = ver;
    input.close();
  });

  input.on("close", () => {
    if (version && /^[0-9]+.[0-9]+.[0-9]+$/.test(version)) {
      console.log("\n准备构建版本: ", version);

      packageFile.version = version;
      fs.writeFileSync(
        path.resolve(process.cwd(), "./package.json"),
        JSON.stringify(packageFile, null, 2)
      );
      fs.readFile(
        path.resolve(process.cwd(), "./CHANGELOG.md"),
        { encoding: "utf-8" },
        (err, data) => {
          let changeLogFile = data;
          if (err) {
            changeLogFile = `# 更新日志${EOL}${EOL}`;
            fs.writeFileSync(
              path.resolve(process.cwd(), "./CHANGELOG.md"),
              changeLogFile,
              { encoding: "utf-8" }
            );
          }
          const changeLogTemplate =
            `# 更新日志${EOL}${EOL}` +
            `## [${version}] - ${TODAY}${EOL}${EOL}` +
            `### Added${EOL}${EOL}- 这里记录新增加了哪些功能／接口 ${EOL}${EOL}` +
            `### Changed${EOL}${EOL}- 功能／接口变更 ${EOL}${EOL}` +
            `### Deprecated${EOL}${EOL}- 不建议使用的功能／接口，将来会删掉 ${EOL}${EOL}` +
            `### Removed${EOL}${EOL}- 之前不建议使用的功能／接口，这次真的删掉了 ${EOL}${EOL}` +
            `### Fixed${EOL}${EOL}- 这里记录解决了哪些问题 ${EOL}${EOL}` +
            `### Others${EOL}${EOL}- 这里记录性能优化和安全性增强等改进 ${EOL}${EOL}`;
          if (changeLogFile.includes(`[${version}]`)) {
            console.log("\n版本号已存在");
          } else if (/# 更新日志/.test(changeLogFile)) {
            changeLogFile = changeLogFile.replace(
              `# 更新日志${EOL}${EOL}`,
              changeLogTemplate
            );
          } else {
            changeLogFile = changeLogTemplate + changeLogFile + EOL + EOL;
          }
          fs.writeFileSync(
            path.resolve(process.cwd(), "./CHANGELOG.md"),
            changeLogFile,
            { encoding: "utf-8" }
          );
          console.log(`\n开始编译版本${version}...`);
        }
      );
    } else {
      throw Error("请输入正确版本号，格式：Major.Minor.Patch");
    }
  });
}

module.exports = main;


src/config.ts
/*
 * @Description: 默认配置
 * @Author: Sunly
 * @Date: 2023-05-23 09:18:29
 */
export type IConfig = {
  title: string;
  checkVersion: boolean;
  changelogTemplate: Record<string, string>[];
};

export const defaultConfig: IConfig = {
  title: "CHANGELOG",
  checkVersion: true,
  changelogTemplate: [
    {
      Feat: "Records what features/interfaces have been added",
    },
    {
      Change: "Records changes in features/interfaces",
    },
    {
      Fix: "Records what issues have been resolved",
    },
    {
      Deprecated:
        "Features/interfaces that are not recommended and will be removed in the future",
    },
    {
      Remove:
        "Features/interfaces that were previously not recommended and are now truly removed",
    },
    {
      Style: "Records changes in style",
    },
    {
      Type: "Records changes in type",
    },
    {
      Other: "Records performance optimizations and security enhancements",
    },
  ],
};


rollup.config.js
/*
 * @Description:
 * @Author: Sunly
 * @Date: 2023-08-16 06:54:50
 */
import { resolve } from "path";
import cleanup from "rollup-plugin-cleanup";
import dts from "rollup-plugin-dts";
import { terser } from "rollup-plugin-terser";
import addSheBang from "rollup-plugin-add-shebang";
import typescript from "@rollup/plugin-typescript";

export default [
  {
    input: resolve(process.cwd(), "./src/index.ts"),
    output: [
      {
        file: resolve(process.cwd(), "./dist/index.cjs"),
        format: "cjs",
      },
      {
        file: resolve(process.cwd(), "./dist/index.mjs"),
        format: "esm",
      },
    ],
    plugins: [typescript(), terser(), cleanup()],
  },
  {
    input: resolve(process.cwd(), "./src/cli.ts"),
    output: [
      {
        file: resolve(process.cwd(), "./dist/cli.js"),
        format: "esm",
      },
    ],
    plugins: [
      typescript(),
      terser(),
      cleanup(),
      addSheBang({
        include: "./dist/cli.js",
      }),
    ],
  },
  {
    input: resolve(process.cwd(), "./src/update-version.ts"),
    output: [
      {
        file: resolve(process.cwd(), "./dist/update-version.js"),
        format: "esm",
      },
    ],
    plugins: [
      typescript(),
      terser(),
      cleanup(),
      addSheBang({
        include: "./dist/update-version.js",
      }),
    ],
  },
  {
    input: resolve(process.cwd(), "./src/index.ts"),
    output: [
      {
        file: resolve(process.cwd(), "./dist/index.d.ts"),
      },
    ],
    plugins: [dts()],
  },
];


src/index.ts
/*
 * @Description: 主入口
 * @Author: Sunly
 * @Date: 2023-05-23 02:37:49
 */
import fs from "fs";
import { EOL } from "os";
import path from "path";
import { defaultConfig, type IConfig } from "./config.js";
import input from "@inquirer/input";
import select from "@inquirer/select";

const CHANGELOG_PATH = path.resolve(process.cwd(), "./CHANGELOG.md");
const CONFIG_PATH = path.resolve(process.cwd(), "./gen-version.config.json");
const PACKAGE_PATH = path.resolve(process.cwd(), "./package.json");

function readConfig(): IConfig {
  let userConfig = "{}";
  if (fs.existsSync(CONFIG_PATH)) {
    userConfig = fs.readFileSync(CONFIG_PATH, "utf-8");
  }
  return { ...defaultConfig, ...JSON.parse(userConfig) };
}

function readPackageFile(): string {
  return fs.readFileSync(PACKAGE_PATH, "utf-8");
}

function getCurVersion(fileContent: string): [number, number, string] {
  const startVersion = fileContent.indexOf(`"version":`) + 10;
  const startIndex = fileContent.indexOf(`"`, startVersion) + 1;
  const endIndex = fileContent.indexOf(`"`, startIndex);
  const curVersion = fileContent.slice(startIndex, endIndex);
  console.log("\n当前构建版本：", curVersion);
  return [startIndex, endIndex, curVersion];
}

async function readUserInput(
  curVersion: string,
  checkVersion: boolean
): Promise<string> {
  // function questionAsync(rl: readline.Interface, query: string) {
  //   return new Promise<string>((resolve) => {
  //     rl.question(query, resolve);
  //   });
  // }

  // const rl = readline.createInterface({
  //   input: process.stdin,
  //   output: process.stdout,
  // });
  const major = Number(curVersion.split(".")[0]);
  const minor = Number(curVersion.split(".")[1]);
  const patch = Number(curVersion.split(".")[2]);

  const nextMajorVersion = `${major + 1}.0.0`;
  const nextMinorVersion = `${major}.${minor + 1}.0`;
  const nextPatchVersion = `${major}.${minor}.${patch + 1}`;

  try {
    const selectRes = await select({
      message: "请选择本次构建的版本：",
      choices: [
        {
          name: `不更新版本: ${curVersion}`,
          value: curVersion,
        },
        {
          name: `Patch: ${nextPatchVersion}`,
          value: nextPatchVersion,
        },
        {
          name: `Minor: ${nextMinorVersion}`,
          value: nextMinorVersion,
        },
        {
          name: `Major: ${nextMajorVersion}`,
          value: nextMajorVersion,
        },
        {
          name: "手动输入版本号",
          value: "custom",
        },
      ],
    });

    if (selectRes !== "custom") {
      return selectRes;
    }

    const inputRes = await input({
      message: "请输入本次构建的版本号：",
      validate: (value) => {
        if (!value) {
          return "版本号不能为空";
        }
        if (checkVersion && !/^[0-9]+.[0-9]+.[0-9]+$/.test(value)) {
          return "请输入正确版本号，格式：Major.Minor.Patch";
        }
        return true;
      },
    });

    return inputRes;
  } catch (error: any) {
    throw new Error("输入版本号发生错误 " + error.message);
  }
}

function updatePackageVersion(
  fileContent: string,
  start: number,
  end: number,
  newVersion: string
): void {
  fs.writeFileSync(
    PACKAGE_PATH,
    fileContent.slice(0, start) + newVersion + fileContent.slice(end)
  );
}

function createChangeLogTemplate(
  version: string,
  title: string,
  template: Record<string, string>[]
): string {
  const YEAR = new Date().getFullYear();
  const MONTH = (new Date().getMonth() + 1).toString().padStart(2, "0");
  const DAY = new Date().getDate().toString().padStart(2, "0");
  const TODAY = `${YEAR}-${MONTH}-${DAY}`;

  return (
    `# ${title}${EOL}${EOL}` +
    `## [${version}] - ${TODAY}${EOL}${EOL}` +
    template
      .map((item) => {
        const key = Object.keys(item)[0];
        const value = Object.values(item)[0];
        return `### ${key}${EOL}${EOL}- ${value} ${EOL}`;
      })
      .join(EOL) +
    EOL
  );
}

function updateChangelog(
  newVersion: string,
  title: string,
  template: Record<string, string>[]
): void {
  if (!fs.existsSync(CHANGELOG_PATH)) {
    try {
      fs.writeFileSync(CHANGELOG_PATH, `# ${title}${EOL}${EOL}`, {
        encoding: "utf-8",
      });
    } catch (error: any) {
      throw new Error("创建CHANGELOG.md失败 " + error.message);
    }
  }

  let changelogFile = "";
  try {
    changelogFile = fs.readFileSync(CHANGELOG_PATH, "utf-8");
    if (changelogFile.includes(`[${newVersion}]`)) {
      console.log("CHANGELOG.md中已存在该版本号\n");
    } else {
      const changeLogTemplate = createChangeLogTemplate(
        newVersion,
        title,
        template
      );
      if (changelogFile.startsWith(`# ${title}`)) {
        changelogFile = changelogFile.replace(
          `# ${title}${EOL}${EOL}`,
          changeLogTemplate
        );
      } else {
        changelogFile = changeLogTemplate + changelogFile + EOL + EOL;
      }
    }
  } catch (error: any) {
    throw new Error("更新CHANGELOG.md失败 " + error.message);
  }

  try {
    fs.writeFileSync(CHANGELOG_PATH, changelogFile, { encoding: "utf-8" });
  } catch (error: any) {
    throw new Error("写入CHANGELOG.md失败 " + error.message);
  }
}

async function main() {
  try {
    const { changelogTemplate, checkVersion, title } = readConfig();
    const packageFile = readPackageFile();
    const [replaceStart, replaceEnd, curVersion] = getCurVersion(packageFile);
    const newVersion = await readUserInput(curVersion, checkVersion);
    updatePackageVersion(packageFile, replaceStart, replaceEnd, newVersion);
    updateChangelog(newVersion, title, changelogTemplate);
    console.log("\n版本更新完成");
  } catch (error: any) {
    console.error(error.message);
  }
}

export default main;
